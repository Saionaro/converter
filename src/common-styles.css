@font-face {
  font-family: "JetBrainsMono";
  src: url("../public/fonts/JetBrainsMono-Light.woff2");
  font-weight: 200;
  font-style: normal;
  font-display: swap;
}

@font-face {
  font-family: "JetBrainsMono";
  src: url("../public/fonts/JetBrainsMono-Regular.woff2");
  font-weight: 400;
  font-style: normal;
  font-display: swap;
}

*,
*::before,
*::after {
  box-sizing: border-box;
}

body {
  font-family: JetBrainsMono, monospace;
}

body,
h1,
h2,
h3,
h4,
p,
ul,
ol,
li,
figure,
figcaption,
blockquote,
dl,
dd {
  margin: 0;
}

ul,
ol {
  padding: 0;
}

ul,
ol {
  list-style: none;
}

input,
button,
textarea,
select {
  font: inherit;
}

main {
  display: flex;
  width: 100vw;
  height: 100vh;
  align-items: center;
  justify-content: center;
}

@media (prefers-reduced-motion: reduce) {
  * {
    animation-duration: 0.01ms !important;
    animation-iteration-count: 1 !important;
    transition-duration: 0.01ms !important;
    scroll-behavior: auto !important;
  }
}
