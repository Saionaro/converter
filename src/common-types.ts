import { Currency } from "src/constants/currencies";

export interface Wallet {
  currency: Currency;
  amount: string;
  empty: boolean;
}

export type RatesMap = Record<Currency, number>;

export type { Currency };

export type TransactionComponent = [Currency, string];
export type Transaction = [TransactionComponent, TransactionComponent];
