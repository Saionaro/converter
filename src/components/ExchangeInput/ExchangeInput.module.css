.root {
  transition-duration: 0.3s;
  transition-property: opacity, outline;
}

.disabled {
  opacity: 0.1;
}

.active {
  outline: 1px solid red;
}
