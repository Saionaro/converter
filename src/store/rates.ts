import { StoreonModule } from "storeon";
import { Currency } from "src/common-types";
import { REFRESH_INTERVAL } from "src/constants/rates";

type RatesMap = Record<Currency, string>;

export type RatesStore = { rates: RatesMap };

export interface RatesEvents {
  "rates/set": RatesMap;
}

const defaultState: RatesMap = {
  GBP: "0",
  EUR: "0",
  USD: "0",
};

async function fetchRates(callback: (rates: RatesMap) => void): Promise<void> {
  try {
    // await fetch rates
    callback(defaultState);
  } catch (e) {}
}

function syncRates(setter: (data: RatesMap) => void) {
  fetchRates(setter);
  setInterval(() => {
    fetchRates(setter);
  }, REFRESH_INTERVAL);
}

export const rates: StoreonModule<RatesStore, RatesEvents> = (store) => {
  store.on("@init", () => {
    syncRates((data) => store.dispatch("rates/set", data));

    return { rates: defaultState };
  });
  store.on("rates/set", (_, rates) => ({ rates }));
};
