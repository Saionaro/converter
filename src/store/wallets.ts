import { StoreonModule } from "storeon";
import { Currency, Wallet } from "src/common-types";

type WalletsPartial = Partial<Record<Currency, Wallet>>;

export type WalletsStore = { wallets: WalletsPartial };

export interface WalletsEvents {
  "wallets/set": WalletsPartial;
}

export const wallets: StoreonModule<WalletsStore, WalletsEvents> = (store) => {
  store.on("@init", () => ({}));
  store.on("wallets/set", (oldData, data) => ({
    wallets: { ...oldData.wallets, ...data },
  }));
};
