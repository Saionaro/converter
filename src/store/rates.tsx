import React, {
  createContext,
  useEffect,
  useState,
  useMemo,
  ReactNode,
} from "react";
import { Currency, RatesMap } from "src/common-types";
import { REFRESH_INTERVAL } from "src/constants/rates";
import { CURRENCIES } from "src/constants/currencies";
import { fetchRates } from "src/api/rates";

export type RatesStore = {
  rates: RatesMap;
  error?: string;
};

const initialState = CURRENCIES.reduce((acc, cur: Currency) => {
  acc[cur] = 1;
  return acc;
}, {} as RatesMap);

export const RatesContext = createContext<RatesStore>({ rates: initialState });

type Props = { children: ReactNode };

export const RatesProvider = ({ children }: Props) => {
  const [state, setState] = useState<RatesMap>(initialState);
  const [error, setError] = useState<string | undefined>();
  const value = useMemo(() => ({ rates: state, error }), [state, error]);

  useEffect(() => {
    fetchRates().then(setState).catch(setError);

    const refreshInterval = setInterval(() => {
      // fetchRates().then(setState).catch(setError);
    }, REFRESH_INTERVAL);

    return () => clearInterval(refreshInterval);
  }, []);

  return (
    <RatesContext.Provider value={value}>{children}</RatesContext.Provider>
  );
};
