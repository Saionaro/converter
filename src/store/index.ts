import { useContext } from "react";
import { RatesContext } from "./rates";
import { WalletsContext } from "./wallets";

export function useRates() {
  return useContext(RatesContext);
}

export function useWallets() {
  return useContext(WalletsContext);
}

export { Wrapper } from "./Wrapper";
