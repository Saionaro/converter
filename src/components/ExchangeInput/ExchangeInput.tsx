import React from "react";
import { Currency } from "src/common-types";
import { MoneyValue } from "src/components/MoneyValue";
import { convert } from "src/utils/convert";
import { MoneyInput } from "src/components/MoneyInput";
import { useRates, useWallets } from "src/store";

import st from "./ExchangeInput.module.css";

interface Props {
  currency: Currency;
  pair: Currency;
  value: string;
  onChange: (val: string) => void;
  disabled?: boolean;
  negative?: boolean;
}

export function ExchangeInput({
  currency,
  pair,
  value,
  onChange,
  negative,
  disabled,
}: Props) {
  const { wallets } = useWallets();
  const { rates } = useRates();
  const currentWallet = wallets[currency];
  const pairCost = convert({ from: currency, to: pair, amount: "1", rates });

  return (
    <div className={st.root}>
      <MoneyInput
        value={value}
        currency={currency}
        onChange={onChange}
        negative={negative}
        disabled={disabled}
      />
      <div className={st.info}>
        <span>
          <span>You have </span>
          <MoneyValue currency={currency}>{currentWallet.amount}</MoneyValue>
        </span>
        <span>
          <MoneyValue currency={currency}>1</MoneyValue>
          <span>{" = "}</span>
          <MoneyValue currency={pair}>{pairCost}</MoneyValue>
        </span>
      </div>
    </div>
  );
}
