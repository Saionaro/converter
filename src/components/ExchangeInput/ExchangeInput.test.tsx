import React from "react";
import { render } from "@testing-library/react";
import { Wrapper } from "src/store";
import { ExchangeInput } from "./ExchangeInput";

import "@testing-library/jest-dom/extend-expect";

describe("components/ExchangeInput", () => {
  test("fit snapshot", () => {
    const spy = jest.spyOn(React, "useEffect").mockImplementation();

    const { container } = render(
      <Wrapper>
        <ExchangeInput
          currency="USD"
          pair="GBP"
          value=""
          onChange={jest.fn()}
        />
      </Wrapper>
    );

    expect(container.firstChild).toMatchSnapshot();
    spy.mockRestore();
  });
});
