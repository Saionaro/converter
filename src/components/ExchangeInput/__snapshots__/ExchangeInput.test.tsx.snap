// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ExchangeInput fit snapshot 1`] = `
<div>
  <div
    role="button"
  >
    <input
      data-testid="input"
      inputmode="decimal"
      maxlength="25"
      type="number"
      value=""
    />
    <h2>
      USD
    </h2>
    <div>
      <span
        class=""
      >
        $
      </span>
      <span
        class="undefined"
      >
        0
      </span>
      <span
        class=""
      >
        .
      </span>
      <span
        class=""
      >
        0
      </span>
      <span
        class=""
      >
        0
      </span>
    </div>
  </div>
  <div>
    <span>
      <span>
        You have 
      </span>
      <span>
        $1,500.00
      </span>
    </span>
    <span>
      <span>
        $1.00
      </span>
      <span>
         = 
      </span>
      <span>
        Â£1.00
      </span>
    </span>
  </div>
</div>
`;
